### DISCUSSION TITLE
@Component, @Service, @Controller, @Repository의 차이

### DISCUSSION THREAD

**Discussion by GIVEN53** [2024-woowacourse-study/level-interview#120] [Spring Core]
Component Scan
WebMvcAutoConfiguration -> RequestMappingHandlerMapping -> ... -> isHandler에서 `@Controller` 확인

**Comment by jimi567** [comment]
Spring 프레임워크에서 사용되는 @Component, @Service, @Controller, @Repository 어노테이션은 각기 다른 역할을 수행하는 클래스들을 식별하고 관리하기 위해 사용된다

1. @Component
 + 가장 일반적인 어노테이션 
 + Spring에서 관리해야 할 빈으로 등록하고자 하는 클래스에 사용됩니다.
 + 구체적인 역할이 없는 일반적인 클래스에 사용됩니다.

2. @Service

+ 비즈니스 로직을 처리하는 서비스 계층에 사용됩니다.
+ @Component에 의미가 부여된 형태로 비즈니스 로직을 수행하는 클래스임을 명시

3. @Controller

+ Spring MVC에서 컨트롤러로 사용되는 클래스에 적용됩니다.
+ HTTP 요청에 대한 응답을 반환하는 역할을 하는 클래스임을 명시
+ 주로 presentation 패키지에 사용됩니다.

4. @Repository

+ 데이터 접근 계층(DAO) 또는 Repository 클래스에 사용 됩니다.
+ DB 와 상호 작용을 하는 클래스임을 명시합니다.


--> @Component에 각기 의미와 역할을 부여하는 어노테이션으로 @Service, @Controller, @Repository가 있습니다.

--- 
Spring에서는 @Component, @Service, @Controller, @Repository 모두  @SpringBootApplication 
 의 Component Scan을 통해 자동으로 빈으로 등록 합니다.

**Comment by GIVEN53** [comment]
`@Component`, `@Service`, `@Controller`, `@Repository`는 각각의 클래스를 특정 역할을 수행하는 Spring Bean으로 등록할 때 사용됩니다. 이 어노테이션들은 클래스가 어떤 역할을 하는지를 명시적으로 나타내며, **Spring의 Component Scan 기능을 통해 자동으로 Bean으로 등록**됩니다. 
`@Service`, `@Controller`, `@Repository` 어노테이션은 내부적으로 `@Component` 어노테이션을 사용하고 있습니다.

#### `@Component`
가장 일반적인 형태의 어노테이션으로, 특정 역할에 종속되지 않는 일반적인 Spring Bean을 나타냅니다.
공통 기능을 제공하는 유틸리티 클래스나, 특정 계층에 속하지 않는 일반적인 컴포넌트를 정의할 때 사용됩니다.

#### `@Service`
비즈니스 로직을 수행하는 클래스에 사용되며 서비스 레이어의 Bean을 나타냅니다.

#### `@Controller`
Spring MVC에서 웹 요청을 처리하는 컨트롤러 클래스에 사용되며 프레젠테이션 레이어의 Bean을 나타냅니다.

#### `@Repository`
데이터베이스와의 상호작용을 수행하는 클래스에 사용되며. 데이터 액세스 레이어의 Bean을 나타냅니다.

### 컨트롤러, 레포지토리에 `@Component` 쓰면 안돼요?
Spring 6(Spring Boot 3) 이전 버전에서는 `@Component` + `@RequestMapping`으로도 Bean 및 핸들러로 등록되었습니다. 하지만 Spring 6 이후 부터 `@Controller` 외에는 핸들러로 등록하지 않아 웹 요청을 정상적으로 수행할 수 없습니다.

_RequestMappingHandlerMapping.class_

<img width="549" alt="스크린샷 2024-06-23 오후 10 31 14" src="https://github.com/2024-woowacourse-study/level-interview/assets/101033262/25b1f40e-34b9-46d2-9bab-fb86fbe6072a">

reference : https://github.com/spring-projects/spring-framework/issues/22154

`@Repository`를 사용할 경우 PersistenceExceptionTranslationPostProcessor가 Bean으로 등록되어 있으면 예외를 DataAccessException 하위 클래스로 변환할 수 있습니다.
reference:
https://www.baeldung.com/spring-bean-annotations#repository
https://docs.spring.io/spring-framework/reference/data-access/orm/general.html#orm-exception-translation

또한 명시적인 의미 전달 및 AOP의 PointCut으로 그루핑하는 등 어노테이션에 따라 부가적인 기능을 부여할 수 있기 때문에 의도에 맞게 어노테이션을 사용하는 것이 바람직합니다.
